core/src/execution/datafusion/expressions/scalar_funcs/hex.rs

```diff
@@ -26,24 +26,69 @@ use arrow_schema::DataType;
 use datafusion::logical_expr::ColumnarValue;
 use datafusion_common::{
     cast::{as_binary_array, as_fixed_size_binary_array, as_int64_array},
-    exec_err, DataFusionError,
+    exec_err, DataFusionError, ScalarValue,
 };
+use datafusion_expr::ScalarFunctionImplementation;
 use std::fmt::Write;
 
 fn hex_int64(num: i64) -> String {
     format!("{:X}", num)
 }
 
-fn hex_bytes<T: AsRef<[u8]>>(bytes: T) -> Result<String, std::fmt::Error> {
-    let bytes = bytes.as_ref();
-    let length = bytes.len();
-    let mut hex_string = String::with_capacity(length * 2);
-    for &byte in bytes {
-        write!(&mut hex_string, "{:02X}", byte)?;
+#[inline(always)]
+fn hex_encode<T: AsRef<[u8]>>(data: T, lower_case: bool) -> String {
+    let mut s = String::with_capacity(data.as_ref().len() * 2);
+    if lower_case {
+        for b in data.as_ref() {
+            // Writing to a string never errors, so we can unwrap here.
+            write!(&mut s, "{b:02x}").unwrap();
+        }
+    } else {
+        for b in data.as_ref() {
+            // Writing to a string never errors, so we can unwrap here.
+            write!(&mut s, "{b:02X}").unwrap();
+        }
     }
+    s
+}
+
+#[inline(always)]
+fn hex_strings<T: AsRef<[u8]>>(data: T) -> String {
+    hex_encode(data, true)
+}
+
+#[inline(always)]
+fn hex_bytes<T: AsRef<[u8]>>(bytes: T) -> Result<String, std::fmt::Error> {
+    let hex_string = hex_encode(bytes, false);
     Ok(hex_string)
 }
 
+pub(super) fn wrap_digest_result_as_hex_string(
```

Comment:
A bit of personal opinion, but while this is doing hexing, IMO because it's related to the hash scalar function, it'd be more appropriate to put it there.